Update to latest blend2d and Cython

Zybulon opened the pull request:
This pull request :
- updates the blend2d dependency
- updates the asmjit dependency
- ensures compatibility with latest Cython
- corrects the order of rgba variables in the _get_rgba32_value function
- accelerate graphics by compiling blend2d in Release mode

Commit:
updating submodule asmjit to latest version

Commit:
updating blend2d submodule to latest

Commit:
update to latest blend2d asmjit and cython compatibility

Commit:
correction of rgba order in _get_rgba32_value

Commit:
Update the examples

Commit:
remove dist folder

Commit:
add dist/ to .gitignore

Commit:
compile blend2d in release mode in order to speed up graphics

jwiggins commented on examples/gradient.py:
```diff
@@ -1,8 +1,8 @@
 import blend2d
 import numpy as np
-from skimage.io import imsave
+import matplotlib.pyplot as plt
```
Happy to consider this, but it's out of scope for the proposed change here.

Zybulon replied:
Well matplotlib allow me to visualise the picture automatically. However I agree with you it is totally out of scope.

jwiggins commented on examples/text.py:
```diff
@@ -1,17 +1,17 @@
-    text = u'Hello World!'
-    font = blend2d.Font('/Library/Fonts/Skia.ttf', 50)
-    canvas.fill_text((100, 100), font, text)
-
-    imsave('text.png', array)
+    text = "Hello World!"
+    # font = blend2d.Font("/Library/Fonts/Skia.ttf", 50)
+    # canvas.fill_text((100, 100), font, text)
```
I'm guessing this slipped in? Removing this code turns this example into a no-op.

Zybulon replied:
You are right, this slipped in. However , the font does not exist in my system and so it did not really worked. I will go back to the initial example.

jwiggins commented on setup.py:
```diff
@@ -69,7 +69,7 @@ def build_extensions(self):
             cmd = ['cmake', ext.cmake_lists_dir] + cmake_args
             subprocess.check_call(cmd, cwd=tmpdir)
-            cmd = ['cmake', '--build', '.', '--config', BUILD_TYPE]
+            cmd = ['cmake', '--build', '.','--parallel','8', '--config', BUILD_TYPE]
```
This is a bit too "magic-numbery" — for lack of a better term — for my taste. It's not really relevant to updating the upstream dependencies.

Zybulon replied:
Totally agree with you, I made that for my tests but I should not have pushed it. I will remove it.

jwiggins commented on src/_capi.pxd:
```diff
@@ -112,40 +114,39 @@ cdef extern from "blend2d.h":
-        BL_MATRIX2D_OP_POST_ROTATE
-        BL_MATRIX2D_OP_POST_ROTATE_PT
-        BL_MATRIX2D_OP_POST_TRANSFORM
+        BL_TRANSFORM_OP_RESET
+        BL_TRANSFORM_OP_ASSIGN
```
Should the `enum` be renamed here as well?

jwiggins commented on src/_capi.pxd:
```diff
@@ -319,15 +326,15 @@ cdef extern from "blend2d.h":
-    ctypedef void (* BLDestroyImplFunc)(void* impl, void* destroyData)
+        
+    ctypedef void (* BLDestroyExternalDataFunc)(void* impl, void* externalData, void* userData)
     ctypedef BLResult (* BLPathSinkFunc)(BLPathCore* path, const void* info, void* closure)
 
     # BLArray
     BLResult blArrayInit(BLArrayCore* self, uint32_t arrayTypeId)
     BLResult blArrayDestroy(BLArrayCore* self)
     BLResult blArrayReset(BLArrayCore* self)
-    BLResult blArrayCreateFromData(BLArrayCore* self, void* data, size_t size, size_t capacity, uint32_t dataAccessFlags, BLDestroyImplFunc destroyFunc, void* destroyData)
+    BLResult blArrayCreateFromData(BLArrayCore* self, void* data, size_t size, size_t capacity, uint32_t dataAccessFlags, BLDestroyExternalDataFunc destroyFunc, void* destroyData)
```
It looks like this function doesn't exist anymore in the [commit](https://github.com/blend2d/blend2d/commit/ca5403c1d02b2bc9d2de581e4cb13e5e80f33860) of blend2d referenced in the submodule file?

jwiggins commented on src/_capi.pxd:
```diff
@@ -451,27 +457,50 @@ cdef extern from "blend2d.h":
+#     BLResult blFontAssignWeak(BLFontCore* self, const BLFontCore* other)
+#     bool blFontEquals(const BLFontCore* a, const BLFontCore* b)
+#     BLResult blFontCreateFromFace(BLFontCore* self, const BLFontFaceCore* face, float size)
+#     BLResult blFontShape(const BLFontCore* self, BLGlyphBufferCore* gb)
+#     BLResult blFontMapTextToGlyphs(const BLFontCore* self, BLGlyphBufferCore* gb, BLGlyphMappingState* stateOut)
+#     BLResult blFontPositionGlyphs(const BLFontCore* self, BLGlyphBufferCore* gb, uint32_t positioningFlags)
+#     BLResult blFontApplyKerning(const BLFontCore* self, BLGlyphBufferCore* gb)
+#     BLResult blFontApplyGSub(const BLFontCore* self, BLGlyphBufferCore* gb, size_t index, BLBitWord lookups)
+#     BLResult blFontApplyGPos(const BLFontCore* self, BLGlyphBufferCore* gb, size_t index, BLBitWord lookups)
+#     BLResult blFontGetMatrix(const BLFontCore* self, BLFontMatrix* out)
+#     BLResult blFontGetMetrics(const BLFontCore* self, BLFontMetrics* out)
+#     BLResult blFontGetDesignMetrics(const BLFontCore* self, BLFontDesignMetrics* out)
+#     BLResult blFontGetTextMetrics(const BLFontCore* self, BLGlyphBufferCore* gb, BLTextMetrics* out)
+#     BLResult blFontGetGlyphBounds(const BLFontCore* self, const uint32_t* glyphData, intptr_t glyphAdvance, BLBoxI* out, size_t count)
+#     BLResult blFontGetGlyphAdvances(const BLFontCore* self, const uint32_t* glyphData, intptr_t glyphAdvance, BLGlyphPlacement* out, size_t count)
+#     BLResult blFontGetGlyphOutlines(const BLFontCore* self, uint32_t glyphId, const BLMatrix2D* userMatrix, BLPathCore* out, BLPathSinkFunc sink, void* closure)
+#     BLResult blFontGetGlyphRunOutlines(const BLFontCore* self, const BLGlyphRun* glyphRun, const BLMatrix2D* userMatrix, BLPathCore* out, BLPathSinkFunc sink, void* closure)
+
+    BLResult blFontInit(BLFontCore* self) ;
+    BLResult blFontDestroy(BLFontCore* self) ;
+    BLResult blFontReset(BLFontCore* self) ;
+    BLResult blFontAssignMove(BLFontCore* self, BLFontCore* other) ;
+    BLResult blFontAssignWeak(BLFontCore* self, const BLFontCore* other) ;
+    bool blFontEquals(const BLFontCore* a, const BLFontCore* b) ;
+    BLResult blFontCreateFromFace(BLFontCore* self, const BLFontFaceCore* face, float size) ;
+    BLResult blFontShape(const BLFontCore* self, BLGlyphBufferCore* gb) ;
+    BLResult blFontMapTextToGlyphs(const BLFontCore* self, BLGlyphBufferCore* gb, BLGlyphMappingState* stateOut) ;
+    BLResult blFontPositionGlyphs(const BLFontCore* self, BLGlyphBufferCore* gb) ;
+    BLResult blFontApplyKerning(const BLFontCore* self, BLGlyphBufferCore* gb) ;
+    BLResult blFontApplyGSub(const BLFontCore* self, BLGlyphBufferCore* gb, const BLBitArrayCore* lookups) ;
+    BLResult blFontApplyGPos(const BLFontCore* self, BLGlyphBufferCore* gb, const BLBitArrayCore* lookups) ;
+    BLResult blFontGetMetrics(const BLFontCore* self, BLFontMetrics* out) ;
+    BLResult blFontGetMatrix(const BLFontCore* self, BLFontMatrix* out) ;
+    BLResult blFontGetDesignMetrics(const BLFontCore* self, BLFontDesignMetrics* out) ;
+    BLResult blFontGetTextMetrics(const BLFontCore* self, BLGlyphBufferCore* gb, BLTextMetrics* out) ;
+    BLResult blFontGetGlyphBounds(const BLFontCore* self, const uint32_t* glyphData, intptr_t glyphAdvance, BLBoxI* out, size_t count) ;
+    BLResult blFontGetGlyphAdvances(const BLFontCore* self, const uint32_t* glyphData, intptr_t glyphAdvance, BLGlyphPlacement* out, size_t count) ;
+    BLResult blFontGetGlyphOutlines(const BLFontCore* self, BLGlyphId glyphId, const BLMatrix2D* userTransform, BLPathCore* out, BLPathSinkFunc sink, void* userData) ;
+    BLResult blFontGetGlyphRunOutlines(const BLFontCore* self, const BLGlyphRun* glyphRun, const BLMatrix2D* userTransform, BLPathCore* out, BLPathSinkFunc sink, void* userData) ;
+
```
I'd prefer for the commented code here to be replaced by the new version introduced here. Minus the semicolons.

Zybulon replied:
I ll do that

jwiggins commented on src/_capi.pxd:
```diff
@@ -568,22 +597,40 @@ cdef extern from "blend2d.h":
+#    BLResult blImageConvert(BLImageCore* self, uint32_t format)
+#    bool blImageEquals(const BLImageCore* a, const BLImageCore* b)
+#    BLResult blImageScale(BLImageCore* dst, const BLImageCore* src, const BLSizeI* size, uint32_t filter, const BLImageScaleFilter* options)
+
+
     # BLImage
-    BLResult blImageInit(BLImageCore* self)
-    BLResult blImageInitAs(BLImageCore* self, int w, int h, uint32_t format)
-    BLResult blImageInitAsFromData(BLImageCore* self, int w, int h, uint32_t format, void* pixelData, intptr_t stride, BLDestroyImplFunc destroyFunc, void* destroyData)
-    BLResult blImageDestroy(BLImageCore* self)
-    BLResult blImageReset(BLImageCore* self)
-    BLResult blImageAssignMove(BLImageCore* self, BLImageCore* other)
-    BLResult blImageAssignWeak(BLImageCore* self, const BLImageCore* other)
-    BLResult blImageAssignDeep(BLImageCore* self, const BLImageCore* other)
-    BLResult blImageCreate(BLImageCore* self, int w, int h, uint32_t format)
-    BLResult blImageCreateFromData(BLImageCore* self, int w, int h, uint32_t format, void* pixelData, intptr_t stride, BLDestroyImplFunc destroyFunc, void* destroyData)
-    BLResult blImageGetData(const BLImageCore* self, BLImageData* dataOut)
-    BLResult blImageMakeMutable(BLImageCore* self, BLImageData* dataOut)
-    BLResult blImageConvert(BLImageCore* self, uint32_t format)
-    bool blImageEquals(const BLImageCore* a, const BLImageCore* b)
-    BLResult blImageScale(BLImageCore* dst, const BLImageCore* src, const BLSizeI* size, uint32_t filter, const BLImageScaleOptions* options)
+    BLResult blImageInit(BLImageCore* self) ;
+    BLResult blImageInitAs(BLImageCore* self, int w, int h, BLFormat format) ;
+    BLResult blImageInitAsFromData(BLImageCore* self, int w, int h, BLFormat format, void* pixelData, intptr_t stride, BLDataAccessFlags accessFlags, BLDestroyExternalDataFunc destroyFunc, void* userData) ;
+    BLResult blImageDestroy(BLImageCore* self) ;
+    BLResult blImageReset(BLImageCore* self) ;
+    BLResult blImageAssignMove(BLImageCore* self, BLImageCore* other) ;
+    BLResult blImageAssignWeak(BLImageCore* self, const BLImageCore* other) ;
+    BLResult blImageAssignDeep(BLImageCore* self, const BLImageCore* other) ;
+    BLResult blImageCreate(BLImageCore* self, int w, int h, BLFormat format) ;
+    BLResult blImageCreateFromData(BLImageCore* self, int w, int h, BLFormat format, void* pixelData, intptr_t stride, BLDataAccessFlags accessFlags, BLDestroyExternalDataFunc destroyFunc, void* userData) ;
+    BLResult blImageGetData(const BLImageCore* self, BLImageData* dataOut) ;
+    BLResult blImageMakeMutable(BLImageCore* self, BLImageData* dataOut) ;
+    BLResult blImageConvert(BLImageCore* self, BLFormat format) ;
+    bool blImageEquals(const BLImageCore* a, const BLImageCore* b) ;
+    BLResult blImageScale(BLImageCore* dst, const BLImageCore* src, const BLSizeI* size, BLImageScaleFilter filter);
```
Same thing here. I prefer not to have large chunks of this file commented out, more so when there are similar versions of the lines which are not commented out.

jwiggins commented on src/_capi.pyx:
```diff
@@ -40,7 +40,7 @@ cdef uint32_t _get_rgba32_value(color):
         alpha = 255
 
-    return (alpha << 24) | (r << 16) | (g << 8) | b
+    return (alpha << 24) | (b << 16) | (g << 8) | r
```
Now I'm curious if my implementation always had this wrong or if it changed upstream? 😅

Zybulon replied:
I think it was wrong because initially without updating blend2d my blue lines where orange. I must say I was a bit suprised !

jwiggins commented on src/context.pxi:
```diff
@@ -334,8 +333,12 @@ cdef class Context:
     def fill_path(self, Path path):
-        _capi.blContextFillPathD(&self._self, &path._self)
-
+        cdef:
+            _capi.BLPoint point
+        point.x = 0.
+        point.y = 0.
+        _capi.blContextFillPathD(&self._self, &point, &path._self)
```
The signature of `fill_path` looks like it might need to change to better match the upstream API?

Zybulon replied:
I thought so but I did not wanted to change the user interface. And also I did not really get the difference with a move_to instruction.

jwiggins reviewed (commented):
I'm more or less on board with the submodule updates here, with caveats.

Commit:
take review into account

Commit:
pass ruff formatter

Zybulon commented:
Hello, 
What about now ?

perara committed:
Update submodules: asmjit to commit 4111cae and blend2d to commit c79b36d

perara committed:
Enhance project structure and dependencies: Added nanobind submodule, updated CMake configuration for nanobind integration, and created build script for wheel generation. Updated .gitignore to exclude build directories.

perara committed:
Update project configuration for cross-compilation: Enhanced pyproject.toml to skip musllinux builds, improved cleanup commands, and specified architectures. Added setup_buildx.sh for Docker buildx setup. Updated setup.cfg with new version and authors. Refined GitHub Actions workflow for multi-architecture builds and publishing to PyPI.

perara committed:
Refine GitHub Actions workflow for wheel building and publishing: Improved conditional logic for architecture builds and publishing steps, ensuring clarity and correctness in the workflow execution.

perara committed:
Refactor GitHub Actions workflow: Removed unnecessary conditional logic for architecture builds in the wheel building process, streamlining the workflow for improved clarity and efficiency.

perara committed:
Enhance GitHub Actions workflow for building wheels: Introduced separate jobs for x86_64 and ARM64 architectures, improved conditional logic for architecture selection, and updated actions versions for better compatibility. Streamlined environment variable setup and artifact uploads.

perara committed:
Enhance setup.py for ARM64 support: Added specific compiler flags for ARM64 architecture to enable NEON and crypto extensions, improving performance on compatible systems.

perara committed:
Refactor firing effect in ReticleRenderer: Updated pulse effect parameters and logic for improved visual feedback during firing. Adjusted firing flag duration to match pulse duration, enhancing the overall user experience. Removed legacy pulse drawing code for clarity.

perara committed:
Update pyproject.toml for improved build configuration: Changed manylinux image to manylinux_2_28 for better compatibility, specified both x86_64 and aarch64 architectures, and increased CMake build parallel level to 4 for enhanced performance during builds.

perara committed:
Add LICENSE and MANIFEST.in files; update pyproject.toml and setup.cfg for broader compatibility

- Introduced LICENSE file with MIT License terms.
- Added MANIFEST.in to include necessary files in the package distribution.
- Updated pyproject.toml to support Python versions 3.8 to 3.12 and adjusted architecture specifications.
- Enhanced setup.cfg with additional metadata and clarified supported Python versions.
- Created test_wheel_build.sh for testing wheel build configuration.
- Refined GitHub Actions workflow for building wheels across multiple platforms.

perara committed:
Enhance CMake configuration for Python integration: Added support for specifying the Python executable through setup.py, improving cross-compilation capabilities. Adjusted Python find strategy to ensure correct version detection.

perara committed:
Enhance setup.py for macOS architecture detection: Added logic to determine target architecture (arm64 or x86_64) based on environment variables and system checks, improving build compatibility. Updated .gitignore to exclude .history directory.

perara committed:
Refactor CMakeLists.txt to allow nanobind and setup.py to manage output naming: Removed the override of OUTPUT_NAME for the BLEND2DPY_TARGET_NAME, ensuring correct naming conventions are maintained.

perara committed:
Add FIXES_APPLIED.md to document critical issues resolved during multi-platform wheel builds

- Documented four key issues and their fixes related to Python version detection, MSBuild parallel build flags, macOS architecture targeting, and module import failures.
- Updated CMakeLists.txt to ensure correct Python executable usage and module naming conventions.
- Enhanced setup.py for platform-specific build configurations and architecture detection.
- Verified successful builds across multiple platforms, preparing for production PyPI publishing.

perara committed:
Update CMakeLists.txt to set output name for Python extension and improve build logging

- Set the output name of the BLEND2DPY_TARGET_NAME to "_capi" for consistent module imports.
- Adjusted Windows-specific properties to ensure the correct .pyd suffix is used.
- Added debug messages to log the building process and output details for better visibility during builds.

perara committed:
Revert src/CMakeLists.txt to fix Linux/macOS import errors

The removal of OUTPUT_NAME broke module imports. The original
OUTPUT_NAME '_capi' is required for proper module installation.

- Reverted to: OUTPUT_NAME '_capi'
- Added debug messages to track build output
- Linux/Windows should work again
- macOS still under investigation

perara committed:
Disable macOS builds until import issue is resolved

macOS wheels build successfully but fail at import with:
  ModuleNotFoundError: No module named 'blend2d._capi'

Changes:
- Disabled macOS in GitHub Actions workflow
- Added *-macosx* to skip patterns in pyproject.toml
- Created MACOS_DISABLED.md documentation

What works now:
✅ Linux (x86_64, aarch64) - 10 wheels
✅ Windows (AMD64) - 5 wheels
❌ macOS - Disabled (can install from source)

Total: 15 wheels for PyPI release

perara committed:
Update package name to blend2d-python in README and setup configuration

- Changed installation command in README.md to reflect new package name.
- Updated setup.cfg to rename the package from blend2d to blend2d-python.
- Modified GitHub Actions workflow for Trusted Publishing to use the new package name.

Zybulon committed:
Merge pull request #1 from perara/main

Merge nanobind API by perara

Zybulon committed:
Remove support de python 3.8 and 3.9 to ensure the wheels compatibility

Zybulon committed:
update license classifier according to PEP 639

https://peps.python.org/pep-0639/#license-terms

Zybulon committed:
removing non recommended compilation options

fp:fast remove behavior specified by IEEE-754

Zybulon committed:
Remove the "BL" prefix on the classes and enums

The BL prefix is not necessary in Python + it makes the names harder to read

Zybulon committed:
Add some useful interfaces for Python

Zybulon committed:
Make sure the example is working

call to the correct fonctions
correction of the drawing of the star

Zybulon committed:
update minimum python requirement
